plugin-server/src/worker/ingestion/event-pipeline.ts

```diff
@@ -0,0 +1,84 @@
+import { PluginEvent } from '@posthog/plugin-scaffold'
+
+import { runInstrumentedFunction } from '../../main/utils'
+import { Hub } from '../../types'
+import { runOnAction, runOnEvent, runOnSnapshot, runProcessEvent } from '../plugins/run'
+import { Action } from './../../types'
+import { ingestEvent } from './ingest-event'
+
+export async function runEventPipeline(server: Hub, event: PluginEvent): Promise<void> {
+    const processedEvent = await processEvent(server, event)
+
+    if (processedEvent) {
+        let actionMatches: Action[] = []
+        await Promise.all([
+            runInstrumentedFunction({
+                server,
+                event: processedEvent,
+                func: async (event) => {
+                    const result = await ingestEvent(server, event)
+                    actionMatches = result.actionMatches || []
+                },
+                statsKey: 'kafka_queue.single_ingestion',
+                timeoutMessage: 'After 30 seconds still ingesting event',
+            }),
+            onEvent(server, processedEvent),
+        ])
+
+        server.statsd?.increment('kafka_queue_single_event_processed_and_ingested')
+
+        if (actionMatches.length > 0) {
+            const promises = []
+            for (const actionMatch of actionMatches) {
+                promises.push(
+                    runInstrumentedFunction({
+                        server,
+                        event: processedEvent,
+                        func: (event) => runOnAction(server, actionMatch, event),
+                        statsKey: `kafka_queue.on_action`,
+                        timeoutMessage: 'After 30 seconds still running onAction',
+                    })
+                )
+            }
+            await Promise.all(promises)
+        }
+    } else {
```

Comment:
given this function has shrunk a bit might make sense to flip this on its head and have:

```js
if (!processedEvent) {
        // processEvent might not return an event. This is expected and plugins, e.g. downsample plugin uses it.
        server.statsd?.increment('kafka_queue.dropped_event', {
            teamID: String(event.team_id),
        })
        return
}
```
and then all the rest outside of the conditional.

Happy to keep as is though if it helps us see that we've just moved code and not changed it

Reply:
See comment above.